File: arrays/04_best_time2buy_sell_stock/test_best_time2buy_sell_stock.py
from unittest import TestCase, TestResult, TestLoader
from best_time2buy_sell_stock import Solution
from time import time

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


@time_tests
class TestBestTimeToBuySellStock(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_OneProfit(self):
        prices = [7, 1, 5, 3, 6, 4]
        expected = 5
        result = self.solution.max_profit(prices)

        self.assertListEqual(result, expected)

    def test2_NoProfit(self):
        prices = [7, 6, 4, 3, 1]
        expected = 0
        result = self.solution.max_profit(prices)

        self.assertListEqual(result, expected)

    def test3_TwoProfits(self):
        prices = [7, 9, 12, 1, 3, 6, 8]
        expected = 7
        result = self.solution.max_profit(prices)

        self.assertListEqual(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestBestTimeToBuySellStock)
    result = CustomTestResult()
    suite.run(result)
    result.printTestTimes()

File: arrays/04_best_time2buy_sell_stock/test_best_time2buy_sell_stock.cpp
#include "best_time2buy_sell_stock.cpp"
#include <gtest/gtest.h>
#include <vector>
#include <chrono>
#include <iostream>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestBestTimeToBuySellStock : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestBestTimeToBuySellStock, test1_OneProfit) { 
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    int expected = 5;
    int result = solution->maxProfit(prices);

    ASSERT_EQ(expected, result);
}

TEST_F(TestBestTimeToBuySellStock, test2_NoProfit) { 
    vector<int> prices = {7, 6, 4, 3, 1};
    int expected = 0;
    int result = solution->maxProfit(prices);

    ASSERT_EQ(expected, result);
}

TEST_F(TestBestTimeToBuySellStock, test3_TwoProfits) { 
    vector<int> prices = {7, 9, 12, 1, 3, 6, 8};
    int expected = 7;
    int result = solution->maxProfit(prices);

    ASSERT_EQ(expected, result);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}